feat(text-input): Add InputGroup component

NicholasBoll opened the pull request:
## Summary

Add an `InputGroup` component to handle icons in a `TextInput` component.

There is already an `InputIconContainer` component that is not documented, but is limited to only right-aligned icons and doesn't handle RTL.

## Release Category
Components

### Release Note
`InputGroup` will replace `InputIconContainer`. `InputIconContainer` does not handle bidirectionality or icons at the start of an input. `InputIconContainer` will be deprecated and later removed in future versions.

Before:
```tsx
<InputIconContainer icon={<SystemIcon icon={exclamationCircleIcon} />} />
```

After
```tsx
<InputGroup>
  <InputGroup.Input />
  <InputGroup.InnerEnd>
    <SystemIcon icon={exclamationCircleIcon} />
  </InputGroup.InnerEnd>
</InputGroup>
```

---

## Checklist

- [x] MDX documentation adheres to Canvas Kit's [standard MDX template](https://github.com/Workday/canvas-kit/discussions/1131)
- [x] Label `ready for review` has been added to PR

## For the Reviewer

<!-- Provide a bit of context about what this PR does. Add any additional checklist items you'd like the reviewer to check -->

- [x] PR title is short and descriptive
- [x] PR summary describes the change (Fixes/Resolves linked correctly)
- [x] PR Release Notes describes additional information useful to call out in a release message or removed if not applicable

## Where Should the Reviewer Start?

<!-- If you were reviewing this PR, where would you want to start?  -->
<!-- e.g. `/modules/react/common/lib/utils/someUtil.ts`  -->

## Areas for Feedback? (optional)

<!-- Do you have any particular areas where you'd like additional focus or feedback from reviewers? -->

- [x] Code
- [x] Documentation
- [x] Testing

## Screenshots or GIFs (if applicable)

<img width="309" alt="image" src="https://user-images.githubusercontent.com/338257/235381870-9e097149-8088-45a6-8017-66ea08b3292e.png">

NicholasBoll commented on modules/react/common/lib/utils/dispatchInputEvent.ts:
```diff
@@ -0,0 +1,16 @@
+export function dispatchInputEvent<T extends HTMLElement>(input: T | null, value: string): void {
+  // Changing value prop programmatically doesn't fire an Synthetic event or trigger native onChange.
+  // We can not just update .value= in setState because React library overrides input value setter
+  // but we can call the function directly on the input as context.
+  // This will cause onChange events to fire no matter how value is updated.
+  if (input) {
+    const nativeInputValue = Object.getOwnPropertyDescriptor(Object.getPrototypeOf(input), 'value');
+    if (nativeInputValue && nativeInputValue.set) {
+      nativeInputValue.set.call(input, value);
+    }
+
+    const event = new Event('input', {bubbles: true});
+
+    input.dispatchEvent(event);
+  }
+}
```
This seemed like a useful function for sending change info to an input. It could go in the `text-input` package, but it is useful for any input-like element.

NicholasBoll commented on modules/react/form-field/stories/visual-testing/stories_TextInput.tsx:
```diff
@@ -76,3 +80,44 @@ TextInputThemedStates.parameters = {
   },
 };
+
+export const InputGroupStates = () => (
```
RTL example:
 https://5d854c26ba934e0020f5e98a-wgbwrnfmtx.chromatic.com/?path=/story/testing-inputs-text-input--input-group-states
 
 
<img width="594" alt="image" src="https://user-images.githubusercontent.com/338257/235489134-faab59c4-2e46-4a5c-a4b1-e6cc9a527a8f.png">

NicholasBoll replied:
Updated:

<img width="597" alt="image" src="https://user-images.githubusercontent.com/338257/235543301-8a0c2fbe-47a8-49be-bf52-e81ec52b571b.png">

alanbsmith commented on modules/react/text-input/spec/InputGroup.spec.tsx:
```diff
@@ -0,0 +1,28 @@
+          </InputGroup.Start>
+          <InputGroup.Input />
+          <InputGroup.End>
+            <TertiaryButton role="presentation" icon={xSmallIcon} size="small" tabIndex={-1} />
+          </InputGroup.End>
```
Seems like adding a `tabIndex={-1}` is going to be an important caveat for teams.

NicholasBoll replied:
I was thinking about introducing a `InputGroup.ClearButton` that has this built in, including clearing the input when clicked.

We don't have specification for it yet. The Preview combobox has this button as focusable. I've also seen this be a non-focusable button because there's already keyboard functionality to clear an input. In my opinion, it should not be focusable via the tab keys. Our accessibility SMEs tend to agree on less focus stops of non-important elements. It makes the interface less predictable.

alanbsmith replied:
I think this is fine for now. Just seems like an area we could improve in the future.

alanbsmith reviewed (changes requested):
A few nits, but overall, this looks good. Thanks, Nicholas.

NicholasBoll commented on modules/react/text-input/lib/InputGroup.tsx:
```diff
@@ -0,0 +1,137 @@
+
+    React.Children.forEach(children, child => {
+      if (React.isValidElement<any>(child) && child.type === InputGroupStart) {
+        paddingInlineStart = (paddingInlineStart || 0) + (child.props.width || 40);
```
Do multiple layers of `calc` work?

NicholasBoll replied:
`paddingInlineStart` is a number. I'll try out these changes locally to make sure everything works.

> Should we set the default for the width prop to space.xl here? Would that save you a lot of || logic?

No, because `child.props` is what's being passed to the child, not the default props of the child, which is why the `|| 40` is repeated here.

NicholasBoll replied:
Multiple layers of `calc` work: `calc(calc(0px + 40px) + 40px);`

NicholasBoll commented on modules/react/text-input/stories/examples/Icons.tsx:
```diff
@@ -0,0 +1,51 @@
+          size="small"
+          tabIndex={-1}
+          transition="opacity 300ms ease"
+          style={{opacity: value ? 1 : 0}}
```
I don't think we should encourage a style prop for styles that change during the lifecycle of the component. Changing the `opacity` will trigger a style recalculation that the browser cannot predictably cache the result.

Here's a performance profile of using the `style` attribute:
<img width="826" alt="image" src="https://user-images.githubusercontent.com/338257/235498759-9cc018df-f6ff-49ce-aa44-3a08844e592b.png">


Here's a performance profile of using the `opacity` style attribute (uses Emotion):
<img width="843" alt="image" src="https://user-images.githubusercontent.com/338257/235499003-7820f4fb-845f-4704-8fa0-4fed2c1e2ed1.png">

The notable difference is the "Recalculate Style". The  `style` attribute has a time of 0.18ms while the `opacity` style prop has a time of 2.24ms. The latter timing will depend on the amount of elements and style selectors on the page. This example only has a single element + Storybook styling and already the difference is significant. I saw an example of the Workday Sheets product took 80ms of style recalc.

NicholasBoll commented on modules/react/text-input/lib/InputGroup.tsx:
```diff
@@ -0,0 +1,137 @@
+});
+
+/**
+ * An `InputGroup` is an Input with optional start and end elements. The start and end elements are
```
I think it should be `TextInput`, but the wording could be better.

NicholasBoll commented:
I've updated to use `calc` based on @alanbsmith's suggestions. I've simplified the logic and added an extra visual test to ensure multiple inline elements are allowed.

<img width="597" alt="image" src="https://user-images.githubusercontent.com/338257/235542490-f9ce7c07-abfe-433f-ba7c-78a37e80a5bc.png">

NicholasBoll commented on modules/react/text-input/lib/InputGroup.tsx:
```diff
@@ -0,0 +1,137 @@
+
+import {createComponent, ExtractProps, useIsRTL} from '@workday/canvas-kit-react/common';
+import {Flex} from '@workday/canvas-kit-react/layout';
+
```
It is added at the top now.

alanbsmith commented on modules/react/text-input/lib/InputIconContainer.tsx:
```diff
@@ -19,6 +19,7 @@ const IconContainer = styled('div')({
   right: space.xxs,
 });
 
+/** @deprecated Use `InputGroup` instead. */
```
Remind me to put something about this in the release guide, so My doesn't have a surprise. 😄

NicholasBoll replied:
I'll deprecate in 9.1 instead and not change v9 comms

alanbsmith commented on modules/react/text-input/stories/examples/Icons.tsx:
```diff
@@ -0,0 +1,143 @@
+          size="small"
+          tabIndex={-1}
+          transition="opacity 300ms ease"
+          opacity={value ? 1 : 1}
```
Did you decide not to use the `style` attr here? And I think this should be:
```tsx
opacity={value ? 1 : 0}
```

NicholasBoll replied:
This was my performance test story. I didn't mean to commit it. I tried a style prop, a style attribute, and a data attribute to test timing and demonstrate elements effected.

The browser does optimize over time. For example, the data attribute has 2 elements effected the first time, but caches the result so that the second time only 1 element is effected. The style prop is 14 effected the first time and 5 the second time. As the page changes, these optimizations are flushed, so it is still better to avoid Emotion for state-based style changes.

alanbsmith reviewed (approved):
Appreciate the updates. 👍🏼

mannycarrera4 commented on modules/react/text-input/lib/InputGroup.tsx:
```diff
@@ -0,0 +1,155 @@
+    const offsetsStart: string[] = [];
+    const offsetsEnd: string[] = [];
+
+    React.Children.forEach(children, child => {
```
This is offsetting the padding based on if there's an element at the start or end so that the input has the correct padding/size instead of it being overlapped by the position absolute elements.

NicholasBoll replied:
I added a comment

mannycarrera4 replied:
WWMA

mannycarrera4 commented on modules/react/text-input/lib/InputGroup.tsx:
```diff
@@ -0,0 +1,155 @@
+  },
+  subComponents: {
+    /**
+     * A component to show at the start of the input. The input's padding will be adjusted to not
```
not sure if we want to mention the word "inside" here. Cause start can be outside.

NicholasBoll replied:
I updated

mannycarrera4 commented on modules/react/text-input/stories/TextInput.stories.mdx:
```diff
@@ -81,6 +82,22 @@ Labels for required fields are suffixed by a red asterisk.
 
+### Icons
+
+`IconGroup` is available to add icons to the `TextInput`. Internally, a container `div` element is
```
Did you mean `InputGroup` ?

NicholasBoll replied:
No. An `InputGroup` adds icons "inside" `TextInput` component

mannycarrera4 replied:
where is icongroup as a component?

NicholasBoll replied:
Oh. I thought you meant `TextInput` should be `InputGroup`. You can use the suggestion feature to help narrow it down.
```suggestion
`InputGroup` is available to add icons to the `TextInput`. Internally, a container `div` element is
```

mannycarrera4 commented on modules/react/text-input/stories/examples/Icons.tsx:
```diff
@@ -0,0 +1,143 @@
+          role="presentation"
+          icon={xSmallIcon}
+          size="small"
+          tabIndex={-1}
```
There's some context here as to why we don't want this to be focusable right?

NicholasBoll replied:
I added a comment

mannycarrera4 commented on modules/react/text-input/lib/InputGroup.tsx:
```diff
@@ -0,0 +1,167 @@
+
+/**
+ * An `InputGroup` is a container around a {@link TextInput} with optional start and end elements.
+ * The start and end elements are usually icons or icon buttons. The `InputGroup` will add padding
```
```suggestion
 * The start and end elements are usually icons or icon buttons within the input. The `InputGroup` will add padding
```

NicholasBoll commented on modules/react/text-input/stories/examples/Icons.tsx:
```diff
@@ -0,0 +1,143 @@
+  opacity: 0,
+  '[data-has-value] &': {
+    opacity: 1,
+  },
```
You can select a range of lines by holding the shift key to make sure the comment is considered outdated when fixed.

mannycarrera4 commented on modules/react/text-input/lib/InputGroup.tsx:
```diff
@@ -0,0 +1,167 @@
+};
+
+/**
+ * An `InputGroup` is a container around a {@link TextInput} with optional start and end elements.
```
```suggestion
 * An `InputGroup` is a container around a {@link TextInput} with optional inner start and inner end elements.
```

mannycarrera4 replied:
no idea if that suggestion makes sense, but trying to be consistent with the component name

mannycarrera4 commented on modules/react/text-input/stories/TextInput.stories.mdx:
```diff
@@ -81,6 +82,22 @@ Labels for required fields are suffixed by a red asterisk.
+### Icons
+
+`InputGroup` is available to add icons to the `TextInput`. Internally, a container `div` element is
+used with relative position styling on the div and absolute position styling on the start and end
```
```suggestion
used with relative position styling on the `div` and absolute position styling on the start and end
```

mannycarrera4 commented on modules/react/text-input/stories/TextInput.stories.mdx:
```diff
@@ -81,6 +82,22 @@ Labels for required fields are suffixed by a red asterisk.
+[CSS Logical Properties](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Logical_Properties)
+and will be semantically correct in left-to-right and right-to-left languages.
+
+`InputGroup.Start` and `InputGroup.End` subcomponents can handle any child elements, but are built
```
```suggestion
`InputGroup.InnerStart` and `InputGroup.InnerEnd` subcomponents can handle any child elements, but are built
```

mannycarrera4 commented on modules/react/text-input/stories/TextInput.stories.mdx:
```diff
@@ -81,6 +82,22 @@ Labels for required fields are suffixed by a red asterisk.
+
+`InputGroup.Start` and `InputGroup.End` subcomponents can handle any child elements, but are built
+for icons. The default width is `40px`, which is perfect for icons. If you need to use something
+else, be sure to set the `width` property of `InputGroup.Start` or `InputGroup.End` to match the
```
```suggestion
else, be sure to set the `width` property of `InputGroup.InnerStart` or `InputGroup.InnerEnd` to match the
```

NicholasBoll replied:
Thanks for catching that.
